pkg/actuators/machineset/ec2_instance_types.go

```diff
@@ -13,3763 +13,166 @@ limitations under the License.
-	},
-	"z1d.xlarge": {
-		InstanceType: "z1d.xlarge",
-		VCPU:         4,
-		MemoryMb:     32768,
-		GPU:          0,
-		Architecture: "amd64",
-	},
+// InstanceTypesCache is a cache for instance type information.
+type InstanceTypesCache interface {
+	GetInstanceType(awsClient awsclient.Client, cacheID string, instanceType string) (InstanceType, error)
+}
+
+// instanceTypesRegion holds cached instance types for specific region and time when it was last updated.
+type instanceTypesRegion struct {
+	instanceTypes map[string]InstanceType
+	lastUpdate    time.Time
+}
+
+// instanceTypesCache holds cached instance types per region. Acess is synchronized via rwmutex.
+type instanceTypesCache struct {
+	cache   map[string]instanceTypesRegion
+	rwmutex sync.RWMutex
+}
+
+// NewInstanceTypesCache creates an empty instance types cache.
+func NewInstanceTypesCache() InstanceTypesCache {
+	cache := &instanceTypesCache{}
+	cache.cache = map[string]instanceTypesRegion{}
+	cache.rwmutex = sync.RWMutex{}
+	return cache
+}
+
+// GetInstanceType retrievees InstanceType from cache by name. If the cache is stale or nil it is refreshed first from the EC2 API.
+// The fetched instance types are specific to the region of the awsClient. Using region name as cacheID is recomended.
+func (i *instanceTypesCache) GetInstanceType(awsClient awsclient.Client, cacheID string, instanceType string) (InstanceType, error) {
+	i.rwmutex.RLock()
+
+	if !i.isCacheFresh(cacheID) {
+		i.rwmutex.RUnlock()
+		if err := i.refresh(awsClient, cacheID); err != nil {
+			return InstanceType{}, fmt.Errorf("error refreshing instance types cache: %w", err)
+		}
+		i.rwmutex.RLock()
+	}
+
+	instanceTypeInfo, ok := i.cache[cacheID].instanceTypes[instanceType]
+	if !ok {
+		instanceNames := []string{}
+		for _, instanceType := range i.cache[cacheID].instanceTypes {
+			instanceNames = append(instanceNames, instanceType.InstanceType)
+		}
+		return InstanceType{}, fmt.Errorf("instance type %q not found: The valid instance types in the current region are: %q", instanceType, instanceNames)
+	}
+
+	i.rwmutex.RUnlock()
+	return instanceTypeInfo, nil
+}
+
+// isCacheFresh checks whether the cache for given cacheId is populated and has been refreshed in the last 24 hours.
+func (i *instanceTypesCache) isCacheFresh(cacheID string) bool {
+	cacheForRegion, ok := i.cache[cacheID]
+	return ok && cacheForRegion.instanceTypes != nil && cacheForRegion.lastUpdate.After(time.Now().Add(-24*time.Hour))
+}
+
+// refresh ensures that the cache is updated in a thread safe way.
+func (i *instanceTypesCache) refresh(awsClient awsclient.Client, cacheID string) error {
+	// Only one thread should refresh the cache at a time.
+	// Parallel refresh does not speed up the process and can cause throttling.
+	i.rwmutex.Lock()
+	defer i.rwmutex.Unlock()
+	if i.isCacheFresh(cacheID) {
+		// Another thread has already refreshed the cache.
+		return nil
+	}
+	instanceTypes, err := fetchEC2InstanceTypes(awsClient)
```

Comment:
A little bit more whitespace would help here to break up each of the if statements